raft.go

```diff
@@ -520,100 +510,221 @@ func (r *Raft) leaderLoop() {
 
 		case <-lease:
 			// Check if we've exceeded the lease, potentially stepping down
-			maxDiff := r.checkLeaderLease()
+			r.checkLeaderLease()
+			// Renew the lease timer
+			lease = time.After(r.conf.LeaderLeaseTimeout)
 
-			// Next check interval should adjust for the last node we've
-			// contacted, without going negative
-			checkInterval := r.conf.LeaderLeaseTimeout - maxDiff
-			if checkInterval < minCheckInterval {
-				checkInterval = minCheckInterval
+		case progress := <-r.peerProgressCh:
+			peer, ok := r.peers[progress.peerID]
+			if ok {
+				peer.progress = progress
+				r.computeLeaderProgress()
 			}
 
-			// Renew the lease timer
-			lease = time.After(checkInterval)
-
 		case <-r.shutdownCh:
 			return
 		}
+
 	}
 }
 
-// verifyLeader must be called from the main thread for safety.
-// Causes the followers to attempt an immediate heartbeat.
-func (r *Raft) verifyLeader(v *verifyFuture) {
-	// Current leader always votes for self
-	v.votes = 1
-
-	// Set the quorum size, hot-path for single node
-	v.quorumSize = r.quorumSize()
-	if v.quorumSize == 1 {
-		v.respond(nil)
-		return
+func (r *Raft) updateCommitIndex(oldCommitIndex, commitIndex uint64) {
+	stepDown := false
+	r.setCommitIndex(commitIndex)
+	// Process the newly committed entries
+	if r.configurations.latestIndex > oldCommitIndex &&
+		r.configurations.latestIndex <= commitIndex {
+		r.configurations.committed = r.configurations.latest
+		r.configurations.committedIndex = r.configurations.latestIndex
+		if !hasVote(r.configurations.committed, r.localID) {
+			stepDown = true
+		}
 	}
 
-	// Track this request
-	v.notifyCh = r.verifyCh
-	r.leaderState.notify[v] = struct{}{}
+	for {
+		e := r.leaderState.inflight.Front()
+		if e == nil {
+			break
+		}
+		commitLog := e.Value.(*logFuture)
+		idx := commitLog.log.Index
+		if idx > commitIndex {
+			break
+		}
+		// Measure the commit time
+		metrics.MeasureSince([]string{"raft", "commitTime"}, commitLog.dispatch)
+		r.processLogs(idx, commitLog)
+		r.leaderState.inflight.Remove(e)
+	}
 
-	// Trigger immediate heartbeats
-	for _, repl := range r.leaderState.replState {
-		repl.notifyLock.Lock()
-		repl.notify = append(repl.notify, v)
-		repl.notifyLock.Unlock()
-		asyncNotifyCh(repl.notifyCh)
+	if stepDown {
+		if r.conf.ShutdownOnRemove {
+			r.logger.Printf("[INFO] raft: Removed ourself, shutting down")
+			r.Shutdown()
+		} else {
+			r.logger.Printf("[INFO] raft: Removed ourself, transitioning to follower")
+			r.stepDown()
+		}
 	}
+
+	r.updatePeers()
 }
 
-// checkLeaderLease is used to check if we can contact a quorum of nodes
-// within the last leader lease interval. If not, we need to step down,
-// as we may have lost connectivity. Returns the maximum duration without
-// contact. This must only be called from the main thread.
-func (r *Raft) checkLeaderLease() time.Duration {
-	// Track contacted nodes, we can always contact ourself
-	contacted := 1
+func (r *Raft) verified(verifiedCounter uint64) {
```

Comment:
// Responds to any verify futures that have been satisfied. A majority of the voting servers in the cluster have acknowledged this server's leadership since the given count.